File: hello-java/audio-stream-server/src/main/java/org/feuyeux/mmap/StreamManager.java
package org.feuyeux.mmap;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReentrantLock;

public class StreamManager {
    private final Config config;
    private final Map<String, Stream> streams = new ConcurrentHashMap<>();
    private final Map<String, ReentrantLock> locks = new ConcurrentHashMap<>();
    private final ReentrantLock regLock = new ReentrantLock();

    public record Config(
        String cacheDir,
        long maxIdleHours,
        long maxUploadingHours,
        long maxStreams
    ) {
        public static Config defaults() {
            return new Config(
                System.getProperty("cache.dir", "cache"),
                24, 1, 1000
            );
        }
    }

    public record Stats(
        int totalStreams,
        long uploadingStreams,
        long readyStreams,
        long errorStreams
    ) {}

    public static class Stream {
        public final String id;
        public final String path;
        public final MmapCache cache;
        public long off = 0;
        public final Instant created = Instant.now();
        private final java.util.concurrent.atomic.AtomicReference<Instant> lastAccess =
            new java.util.concurrent.atomic.AtomicReference<>(Instant.now());
        public Status status = Status.UPLOADING;

        public enum Status { UPLOADING, READY, ERROR }

        public Stream(String id, String path) throws IOException {
            this.id = id;
            this.path = path;
            this.cache = new MmapCache(path);
            this.cache.create(0);
        }

        public void write(byte[] data) throws IOException {
            var buf = ByteBuffer.wrap(data);
            int n = cache.write(off, buf);
            if (n > 0) off += n;
        }

        public byte[] read(long off, int len) throws IOException {
            var buf = cache.read(off, len);
            if (buf == null) return null;
            byte[] r = new byte[buf.remaining()];
            buf.get(r);
            return r;
        }

        public void touch() {
            lastAccess.set(Instant.now());
        }

        public Instant getLastAccess() {
            return lastAccess.get();
        }

        public void close() throws IOException {
            cache.close();
        }
    }

    public StreamManager() {
        this(Config.defaults());
    }

    public StreamManager(Config config) {
        this.config = config;
        try {
            var p = Paths.get(config.cacheDir);
            if (!Files.exists(p)) Files.createDirectories(p);
        } catch (IOException e) {
            throw new RuntimeException("Cannot init cache dir", e);
        }
    }

    public boolean create(String id) {
        regLock.lock();
        try {
            if (streams.containsKey(id)) return false;
            if (streams.size() >= config.maxStreams) return false;
            var s = new Stream(id, config.cacheDir + "/" + id + ".cache");
            locks.put(id, new ReentrantLock());
            streams.put(id, s);
            return true;
        } catch (Exception e) {
            return false;
        } finally {
            regLock.unlock();
        }
    }

    public Stream get(String id) {
        return streams.computeIfPresent(id, (k, s) -> {
            s.touch();
            return s;
        });
    }

    public boolean complete(String id) {
        var s = get(id);
        if (s == null || s.status != Stream.Status.UPLOADING) return false;
        var lock = locks.get(id);
        if (lock == null) return false;
        lock.lock();
        try {
            if (s.status == Stream.Status.UPLOADING) {
                s.status = Stream.Status.READY;
                return true;
            }
            return false;
        } finally {
            lock.unlock();
        }
    }

    public void write(String id, byte[] data) {
        var s = get(id);
        if (s == null || s.status != Stream.Status.UPLOADING) return;
        var lock = locks.get(id);
        if (lock == null) return;
        lock.lock();
        try {
            s.write(data);
        } catch (IOException e) {
            s.status = Stream.Status.ERROR;
        } finally {
            lock.unlock();
        }
    }

    public byte[] read(String id, long off, int len) {
        var s = get(id);
        if (s == null) return new byte[0];
        var lock = locks.get(id);
        if (lock == null) return new byte[0];
        lock.lock();
        try {
            return s.read(off, len);
        } catch (IOException e) {
            return new byte[0];
        } finally {
            lock.unlock();
        }
    }

    public void delete(String id) {
        regLock.lock();
        try {
            var s = streams.remove(id);
            if (s != null) {
                try { s.close(); } catch (IOException e) {}
                Files.deleteIfExists(Paths.get(s.path));
            }
            locks.remove(id);
        } catch (IOException e) {
            // ignore
        } finally {
            regLock.unlock();
        }
    }

    public List<String> list() {
        return List.copyOf(streams.keySet());
    }

    public void cleanup() {
        regLock.lock();
        try {
            var now = Instant.now();
            var toDelete = new java.util.ArrayList<String>();
            for (var e : streams.entrySet()) {
                var s = e.getValue();
                var idle = java.time.Duration.between(s.getLastAccess(), now);
                // Clean up streams idle for more than maxIdleHours, or uploading for more than maxUploadingHours
                if (idle.toHours() > config.maxIdleHours ||
                    (s.status == Stream.Status.UPLOADING && idle.toHours() > config.maxUploadingHours)) {
                    toDelete.add(e.getKey());
                }
            }
            for (var id : toDelete) {
                delete(id);
            }
        } finally {
            regLock.unlock();
        }
    }

    public Stats getStats() {
        var values = streams.values();
        return new Stats(
            values.size(),
            values.stream().filter(s -> s.status == Stream.Status.UPLOADING).count(),
            values.stream().filter(s -> s.status == Stream.Status.READY).count(),
            values.stream().filter(s -> s.status == Stream.Status.ERROR).count()
        );
    }
}

File: hello-go/src/server/stream_manager.go
package server

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sync"
	"time"
)

const (
	maxStreams        = 1000
	maxIdleHours      = 24
	maxUploadingHours = 1
)

// StreamStatus represents the lifecycle state of a stream.
type StreamStatus string

const (
	StatusUploading StreamStatus = "UPLOADING"
	StatusReady     StreamStatus = "READY"
	StatusError     StreamStatus = "ERROR"
)

// Stream holds the state for a single audio stream.
type Stream struct {
	id         string
	path       string
	cache      *MmapCache
	offset     int64
	created    time.Time
	lastAccess time.Time
	status     StreamStatus
	mu         sync.Mutex
}

// Stats summarises the current StreamManager state.
type Stats struct {
	Total     int
	Uploading int
	Ready     int
	Error     int
}

// StreamManager manages all active streams.
// All public methods are thread-safe.
type StreamManager struct {
	cacheDir string
	streams  map[string]*Stream
	mu       sync.RWMutex // protects streams map
}

func newStreamManager(cacheDir string) *StreamManager {
	if err := os.MkdirAll(cacheDir, fs.ModePerm); err != nil {
		panic(fmt.Sprintf("cannot create cache dir: %v", err))
	}
	return &StreamManager{
		cacheDir: cacheDir,
		streams:  make(map[string]*Stream),
	}
}

// create registers a new stream and opens its backing mmap cache.
func (sm *StreamManager) create(id string) bool {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if _, exists := sm.streams[id]; exists {
		return false
	}
	if len(sm.streams) >= maxStreams {
		return false
	}
	path := filepath.Join(sm.cacheDir, id+".cache")
	cache := newMmapCache(path)
	if err := cache.create(0); err != nil {
		return false
	}
	sm.streams[id] = &Stream{
		id:         id,
		path:       path,
		cache:      cache,
		created:    time.Now(),
		lastAccess: time.Now(),
		status:     StatusUploading,
	}
	return true
}

// get returns the stream and updates its last-access time.
func (sm *StreamManager) get(id string) *Stream {
	sm.mu.RLock()
	s := sm.streams[id]
	sm.mu.RUnlock()
	if s != nil {
		s.mu.Lock()
		s.lastAccess = time.Now()
		s.mu.Unlock()
	}
	return s
}

// complete transitions a stream from UPLOADING to READY.
func (sm *StreamManager) complete(id string) bool {
	s := sm.get(id)
	if s == nil {
		return false
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.status != StatusUploading {
		return false
	}
	if err := s.cache.finalize(s.offset); err != nil {
		s.status = StatusError
		return false
	}
	s.status = StatusReady
	return true
}

// write appends data to an uploading stream.
func (sm *StreamManager) write(id string, data []byte) {
	s := sm.get(id)
	if s == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.status != StatusUploading {
		return
	}
	n, err := s.cache.write(s.offset, data)
	if err != nil {
		s.status = StatusError
		return
	}
	s.offset += int64(n)
}

// read reads data from a stream at the given offset.
func (sm *StreamManager) read(id string, offset int64, length int) []byte {
	s := sm.get(id)
	if s == nil {
		return nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	data, err := s.cache.read(offset, length)
	if err != nil {
		return nil
	}
	return data
}

// delete removes a stream and its backing file.
func (sm *StreamManager) delete(id string) {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	s, ok := sm.streams[id]
	if !ok {
		return
	}
	delete(sm.streams, id)
	s.cache.close()
	_ = os.Remove(s.path)
}

// markError transitions an UPLOADING stream to ERROR (on disconnect).
func (sm *StreamManager) markError(id string) {
	s := sm.get(id)
	if s == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.status == StatusUploading {
		s.status = StatusError
	}
}

// list returns all known stream IDs.
func (sm *StreamManager) list() []string {
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	ids := make([]string, 0, len(sm.streams))
	for id := range sm.streams {
		ids = append(ids, id)
	}
	return ids
}

// cleanup removes idle and stale streams.
func (sm *StreamManager) cleanup() {
	now := time.Now()
	sm.mu.Lock()
	var toDelete []string
	for id, s := range sm.streams {
		s.mu.Lock()
		idle := now.Sub(s.lastAccess)
		uploading := s.status == StatusUploading
		s.mu.Unlock()
		if idle > maxIdleHours*time.Hour ||
			(uploading && idle > maxUploadingHours*time.Hour) {
			toDelete = append(toDelete, id)
		}
	}
	sm.mu.Unlock()
	for _, id := range toDelete {
		sm.delete(id)
	}
}

// stats returns a summary of stream counts by status.
func (sm *StreamManager) stats() Stats {
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	s := Stats{Total: len(sm.streams)}
	for _, stream := range sm.streams {
		stream.mu.Lock()
		switch stream.status {
		case StatusUploading:
			s.Uploading++
		case StatusReady:
			s.Ready++
		case StatusError:
			s.Error++
		}
		stream.mu.Unlock()
	}
	return s
}
